src/nvim/lua/vim.lua

```diff
@@ -1,2 +1,62 @@
 -- TODO(ZyX-I): Create compatibility layer.
-return {}
+--{{{1 package.path updater function
+-- Last inserted paths. Used to clear out items from package.[c]path when they
+-- are no longer in &runtimepath.
+local last_nvim_paths = {}
+local function _update_package_paths()
+  local new_last_nvim_paths = {}
+  local rtps = vim.api.nvim_list_runtime_paths()
+  for _, key in ipairs({'path', 'cpath'}) do
+    local orig_str = package[key] .. ';'
+    local pathtrails = {}
+    local pathtrails_ordered = {}
+    local orig = {}
+    local orig_paths = {}
+    -- Note: ignores trailing item without trailing `;`. Not using something
+    -- simpler in order to preserve empty items (stand for default path).
+    for s in orig_str:gmatch('[^;]*;') do
+      s = s:sub(1, -2)  -- Strip trailing semicolon
+      orig[#orig + 1] = s
+      orig_paths[s] = true
+      -- Find out path patterns. pathtrail should contain something like
+      -- /?.so, /?/init.lua, /?.lua. This allows not to bother determining what
+      -- correct suffixes are.
+      local pathtrail = s:match('[/\\][^/\\]*%?.*$')
+      if pathtrail and not pathtrails[pathtrail] then
+        pathtrails[pathtrail] = pathtrail:sub(1, 1)
```

Comment:
`pathtrails` main purpose is not a cache, it is a set. I just added some more sense in its existence by turning it into a cache.